File: CrackingTheCodingInterview/TechnicalQuestions/3Sum.java
public class 3Sum {
    public static void main(String[] args) {
        
    }

    public static List<Integer[]> threeNumberSum(int[] array, int target) {
        // Write your code here.
            Arrays.sort(array); //sort the array first
            List<Integer[]> res = new ArrayList<Integer[]>();
            //the hashmap is the scaling data structure that leaves us with O(N)
            for (int i = 0;i < array.length-2; i++) {
                int left = i+1;
                int right = array.length-1;
                while (left < right) {
                    int currSum = array[i] + array[left] + array[right];
                    if (currSum == target) {
                        Integer[] triplet = {array[i], array[left], array[right]};
                        res.add(triplet);
                        left++;
                        right--;
                    } else if (currSum < target) {
                        left++;
                    } else if (currSum > target) {
                        right--;
                    }
                }
            }
        return res;
      }
}

File: CrackingTheCodingInterview/TechnicalQuestions/3Sum.js
var threeSum = function(nums) {
    nums.sort((a, b) => {
        return a-b;
    });
    let triplets = []; //will be a nested array;
    for (let i= 0; i < nums.length-2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        let left = i+1;
        let right = nums.length-1;
        while (left < right) {
            let currSum = nums[i] + nums[left] + nums[right];
            if (currSum === 0) {
                triplets.push([nums[i], nums[left], nums[right]]);
                while (nums[left] === nums[left+1]) left++;
                while (nums[right] === nums[right+1]) right--;
                left++;
                right--;
            } else if (currSum < 0) {
                left++;
            } else if (currSum > 0) {
                right--;
            }
        }
    }
    return triplets;  
};